wrangler-service/src/main/java/io/cdap/wrangler/service/directive/WorkspaceHandler.java

```diff
@@ -81,17 +85,50 @@
  * V2 endpoints for workspace
  */
 public class WorkspaceHandler extends AbstractDirectiveHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(WorkspaceHandler.class);
+
   private static final Gson GSON =
     new GsonBuilder().registerTypeAdapter(Schema.class, new SchemaTypeAdapter()).create();
-
-  private WorkspaceStore store;
+  private WorkspaceStore wsStore;
   private ConnectionDiscoverer discoverer;
 
   @Override
   public void initialize(SystemHttpServiceContext context) throws Exception {
     super.initialize(context);
-    store = new WorkspaceStore(context);
+    wsStore = new WorkspaceStore(context);
     discoverer = new ConnectionDiscoverer(context);
+
+    UpgradeStore upgradeStore = new UpgradeStore(context);
+    if (upgradeStore.isUpgradeComplete()) {
```

Comment:
initialize gets called for every thread created by the handler, and will only get called once a request is made to the handler. It is entirely possible for the upgrade logic to run multiple times with this logic.

I thought the plan was to add an initialize method to the AbstractService class and do the upgrade there?

Reply:
Have this https://github.com/cdapio/cdap/pull/13487 to have the program life cycle for service, changed this pr to do the upgrade on the first instance of the service